Libraries/LibWeb/CSS/CSSNumericType.cpp

```diff
@@ -164,11 +164,19 @@ Optional<CSSNumericType> CSSNumericType::added_to(CSSNumericType const& other) c
 
             // 1. Provisionally apply the percent hint hint to both type1 and type2.
             auto provisional_type1 = type1;
-            provisional_type1.apply_percent_hint(hint);
             auto provisional_type2 = type2;
-            provisional_type2.apply_percent_hint(hint);
 
-            // 2. If, afterwards, all the entries of type1 with non-zero values are contained in type2
+            // 2. If type1 does not have percent hint hint, apply the percent hint hint to type1.
+            if (!type1.percent_hint().has_value()) {
+                provisional_type1.apply_percent_hint(hint);
+            }
+            
+            //    Vice versa for type2.
+            if (!type2.percent_hint().has_value()) {
+                provisional_type2.apply_percent_hint(hint);
+            }
+
+            // 3. If, afterwards, all the entries of type1 with non-zero values are contained in type2
```

Comment:
Where does that come from? I don't see it in the spec link that you can find above the function.

Reply:
It's phrased a bit differently in the spec, but it's there:
> If type1 has a non-null [percent hint](https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue-percent-hint) hint and type2 doesn’t
>
>    [Apply the percent hint](https://drafts.css-houdini.org/css-typed-om-1/#apply-the-percent-hint) hint to type2.
>
>    Vice versa if type2 has a non-null [percent hint](https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue-percent-hint) and type1 doesn’t.

Reply:
Given the spec doesn't define how to "apply the percent hint" to a type that already has a percent hint should this be considered (and reported as) a spec bug?

Reply:
Either this is a spec bug, either we aren't supposed to reach that branch with a type with a percent hint - but we do because we did something wrong. I checked the whole function and AFAICT it's implemented correctly, so I guess it's indeed a spec bug. I hope that @AtkinsSJ has more to say.

If this is a spec bug, we might end up implementing something similar, but definitely not in that shape. We need to explicitly write down the fact that we are not following the spec and a link to an associated spec bug.

Reply:
![image](https://github.com/user-attachments/assets/308a8c05-cd48-4f40-9d71-4016d31db239)

This is step 1 in the attached portion of the spec. While it does not list out the need check the `percent_hint().has_value()` beforehand this is necessary due to the current implementation of `apply_percent_hint` which immediately contains `VERIFY(!percent_hint().has_value());` which fails and crashes. This `VERIFY` does not exist in the spec so another possible pair of solutions would be to undo my changes and simply remove the `VERIFY` which I think to be inefficient or to replace the `VERIFY` with an early out return `if` statement which would probably handle more cases. Thoughts?

Reply:
I understand the logic behind that change, you just _made it work_. But you can't add steps in the code and write them like they were originally in the spec. Otherwise, we won't be following the spec.

It turns out that we were actually implementing the spec incorrectly:
https://discord.com/channels/1247070541085671459/1247090064480014443/1322810519584768070

Reply:
Yes I see what you are saying but the `VERIFY` is not included the in the spec so could we simply remove it? This would solve the crash/bug and still follow the spec.

Reply:
The spec doesn't explicitly include a `VERIFY` but these steps are only valid when applying a hint to a type without a percent hint, hence the `VERIFY`:

> To apply the percent hint hint to a type without a [percent hint](https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue-percent-hint), perform the following steps:

Reply:
Then I suppose we have arrived at a spec bug :)

Reply:
> Given the spec doesn't define how to "apply the percent hint" to a type that already has a percent hint should this be considered (and reported as) a spec bug?

Yeah that needs some kind of spec change, to clarify what we're supposed to do there. Maybe making it a no-op is fine, but the spec should say so. 

@tannerdrake-dev for now, please quote the spec verbatim, and clearly comment where we're going ad-hoc and why. The spec issue should go on https://github.com/w3c/css-houdini-drafts, though it's not been active recently so it might take a little while to get a response. Put a link to your issue in the comment so it's easy to track down later.

eg,

```c++
// 1. Provisionally apply the percent hint hint to both type1 and type2.
// FIXME: Spec doesn't handle calling "apply the percent hint" to a type that already has one.
//        See https://github.com/w3c/css-houdini-drafts/issue/....
//        For now, only apply them to types that don't have them.
```

Reply:
I'm reasonably certain this isn't a spec bug because these "apply the percent hint" calls should never be reached by a type with a present-and-nonzero "percent" exponent and therefore cannot already have a percent hint. This is actually a bug in our code in the condition on line 157.